README.rst
pygi-treeview-dnd
=================

GTK+ provides a `high-level DnD API <https://developer.gnome.org/gtk3/stable/gtk3-GtkTreeView-drag-and-drop.html>`_
that makes implementing drag and drop on TreeView objects (lists, trees, etc)
significantly easier, and far less error prone. There are hundreds of lines of
code in GTK to implement this -- so 

Unfortunately, there's a longstanding bug in PyGObject that prevents python
programs from using this API. `See the bug report for details <https://bugzilla.gnome.org/show_bug.cgi?id=756072>`,
but basically the SelectionData in drag_data_get is a copy instead of the exact
same SelectionData object, and GTK expects you to modify the original object.

This is a workaround that uses some metaclass magic, GTK internals, and ctypes
to patch up custom TreeStore/ListStore objects. It seems to work, and I plan
on using it extensively in Exaile. If it doesn't work well or if there are
issues, then I'll post a note on github.

Ideally, one day PyGObject will have a fix for this issue, and this package
will no longer be necessary.

Installation
============

This project is easily installed via pip:

    pip install pygi-treeview-dnd

Requirements
============

* GTK+ 3.x and PyGObject 3.x
* Currently only tested on Linux, GTK+ 3.20, and Python 2.7

Usage
=====

See `examples` on the github repo for simple examples.

Author
======

Dustin Spicuzza (dustin@virtualroadside.com)

License
=======

LGPL 2.1+ (Same as PyGI)

CHANGELOG.md
0.1.0 - 2016-08-16
------------------
* Initial release, seems to work!


examples/dnd_demo.py
#!/usr/bin/env python
#
# This sample program should be executed like so:
#
#    python treetest.py ~/some/path/*
#
# And it will display the files in the treeview.
#

from __future__ import print_function

# This is only required to make the example with without requiring installation
# - Most of the time, you shouldn't use this hack
import sys
from os.path import join, dirname
sys.path.insert(0, join(dirname(__file__), '..'))

import os.path

import gi
gi.require_version('Gdk', '3.0')
gi.require_version('Gtk', '3.0')

from gi.repository import Gio
from gi.repository import Gdk
from gi.repository import GLib
from gi.repository import Gtk
from gi.repository import GObject

from gi_treeviewdnd import PatchedListStore


class MyTreeView(Gtk.TreeView):
    '''
        TreeView that enables using the high level DnD API
    '''
    
    def __init__(self, args):
        Gtk.TreeView.__init__(self)
        
        self.model = MyTreeViewModel(GObject.TYPE_STRING, GObject.TYPE_STRING)
        self.set_model(self.model)
        
        self.append_column(Gtk.TreeViewColumn('name', Gtk.CellRendererText(), text=0))
        
        targets = [
            ('text/plain', 0, 0)
        ]
        
        self.enable_model_drag_source(Gdk.ModifierType.BUTTON1_MASK,
                                      targets,
                                      Gdk.DragAction.MOVE)
        
        self.enable_model_drag_dest(targets,
                                    Gdk.DragAction.MOVE)
    
        for row in args:
            self.model.append(row)


class MyTreeViewModel(PatchedListStore):
    '''
        Custom ListStore that implements GTK+ high level DnD via the
        TreeDragSource and TreeDragDest interfaces
    '''

    #
    # TreeDragSource
    #
    
    def do_drag_data_get(self, path, selection_data):
        print('MyTreeViewModel.do_drag_data_get', selection_data)
        uri = self[path][1]
        print('- setting uri        ', uri)
        selection_data.set_text(uri, -1)
        return True
    
    def do_row_draggable(self, path):
        print('do_row_draggable', path)
        return True
    
    
    def do_drag_data_delete(self, path):
        print('do_drag_data_delete', path)
        del self[path]
        return True
    
    #
    # TreeDragDest interface
    #
    
    def do_drag_data_received(self, dest_path, selection_data):
        print('MyTreeViewModel.do_drag_data_received')
        received_data = selection_data.get_text()
        
        print('-', received_data)
        
        idx = dest_path.get_indices()[0]
        self.insert(idx, row=(os.path.basename(received_data), received_data))
        
        return True
    
    def do_row_drop_possible(self, dest_path, selection_data):
        print('is possible?')
        return True
    

if __name__ == '__main__':
    
    args = []
    
    #
    # Grab file arguments to put into TreeView
    #
    
    import glob
    
    def do(a):
        args.append((os.path.basename(a), Gio.File.new_for_commandline_arg(a).get_uri()))
    
    for arg in sys.argv[1:]:
        if '*' in arg:
            for a in glob.glob(arg):
                do(a)
        else:
            do(arg)
    
    #
    # Construct demo window
    #
    
    window = Gtk.Window.new(Gtk.WindowType.TOPLEVEL)
    
    sw = Gtk.ScrolledWindow.new()
    tv = MyTreeView(args)
    sw.add(tv)
            
    window.add(sw)
    window.connect('destroy', Gtk.main_quit)
    
    window.set_default_size(400, 300)
    
    window.show_all()
    
    #import sigint
    #with sigint.InterruptibleLoopContext(Gtk.main_quit):
    
    Gtk.main()


gi_treeviewdnd.py
#
# Copyright (C) 2016 Dustin Spicuzza <dustin@virtualroadside.com>
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 2.1 of the License, or (at your option) any later version.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
# USA

import ctypes

import gi
gi.require_version('Gtk', '3.0')
gi.require_version('GIRepository', '2.0')

from gi.types import GObjectMeta

from gi.repository import GIRepository
from gi.repository import Gtk
from gi.repository import GObject

__all__ = ['TreeDragSourceMeta', 'PatchedListStore', 'PatchedTreeStore']

# Load the libraries we need via GIRepository
def _get_shared_library(n):
    repo = GIRepository.Repository.get_default()
    return repo.get_shared_library(n).split(',')[0]

def _fn(dll, name, args, res=None):
    fn = getattr(dll, name)
    fn.restype = res
    fn.argtypes = args
    return fn

_gobject_dll = ctypes.CDLL(_get_shared_library('GObject'))
_gtk_dll = ctypes.CDLL(_get_shared_library('Gtk'))

g_value_set_object =                    _fn(_gobject_dll, 'g_value_set_object',
                                            (ctypes.c_void_p, ctypes.c_void_p))
                                            
g_value_set_static_boxed =              _fn(_gobject_dll, 'g_value_set_static_boxed',
                                            (ctypes.c_void_p, ctypes.c_void_p))

gtk_selection_data_get_data_type =      _fn(_gtk_dll, 'gtk_selection_data_get_data_type',
                                            (ctypes.c_void_p,), ctypes.c_void_p)

gtk_selection_data_set =                _fn(_gtk_dll, 'gtk_selection_data_set',
                                            (ctypes.c_void_p, ctypes.c_void_p, ctypes.c_int, ctypes.c_void_p, ctypes.c_int))

_drag_data_get_func = ctypes.CFUNCTYPE(ctypes.c_bool,
                                       ctypes.c_void_p,
                                       ctypes.c_void_p,
                                       ctypes.c_void_p)

def create_drag_data_get_thunk(cls):
    '''
        Constructs a custom thunk function for each class
    
        This is in a closure so that we can use the class to construct a
        GObject.Value for converting the widget object
    '''
    
    def _drag_data_get(raw_widget, raw_path, raw_selection_data):
        '''
            Static function that gets inserted as the drag_data_get vfunction,
            instead of using the vfunction implementation provided by pygobject
            
            This function exists so that we can modify the raw selection data
            explicitly, which allows the high-level DnD API to work.
        '''
        
        # It turns out that GValue is really useful for converting raw pointers
        # to/from python objects. We use it + ctypes to xform the values
        
        # Grab the python wrapper for the widget instance
        gv = GObject.Value(cls)
        g_value_set_object(hash(gv), raw_widget)
        widget = gv.get_object()
        
        # Convert the path to a python object via GValue
        v1 = GObject.Value(Gtk.TreePath)
        g_value_set_static_boxed(ctypes.c_void_p(hash(v1)), raw_path)
        path = v1.get_boxed()

        # Convert the selection data too
        v2 = GObject.Value(Gtk.SelectionData)
        g_value_set_static_boxed(ctypes.c_void_p(hash(v2)), raw_selection_data)
        selection_data = v2.get_boxed()
        
        # Call the original virtual function with the converted arguments
        retval = widget.do_drag_data_get(path, selection_data)
        
        # At this point, selection_data has the information, but it's still just
        # a copy. Copy the data back to the original selection data
        data = selection_data.get_data()
        gtk_selection_data_set(raw_selection_data,
                               gtk_selection_data_get_data_type(hash(selection_data)),
                               selection_data.get_format(),
                               data, len(data))
        
        return retval

    return _drag_data_get_func(_drag_data_get)


def vfunc_info_get_address(vfunc_info, gtype):
    '''
        GIRepository.vfunc_info_get_address almost does what we need... but it
        derefs the address, so it cannot be used for our purposes. This code is
        a translation of that C code into python
    '''
    container_info = vfunc_info.get_container()
    if container_info.get_type() == GIRepository.InfoType.OBJECT:
        object_info = container_info
        interface_info = None
        struct_info = GIRepository.object_info_get_class_struct(object_info)
    else:
        interface_info = container_info
        object_info = None
        struct_info = GIRepository.interface_info_get_iface_struct(interface_info)
        
    field_info = GIRepository.struct_info_find_field(struct_info, vfunc_info.get_name())
    if field_info is None:
        raise AttributeError("Could not find struct field for vfunc")
    
    implementor_class = GObject.type_class_ref(gtype)
    if object_info:
        implementor_vtable = implementor_class
    else:
        interface_type = GIRepository.registered_type_info_get_g_type(interface_info)
        implementor_vtable = GObject.type_interface_peek(implementor_class, interface_type)
    
    offset = GIRepository.field_info_get_offset(field_info)
    return hash(implementor_vtable) + offset


class TreeDragSourceMeta(GObjectMeta):
    '''
        Metaclass you can use to patch TreeDragSource.drag_data_get so that it
        works correctly in a python program.
        
        .. note:: Most users will prefer to use PatchedListStore and
                  PatchedTreeStore instead
    '''
        
    def __init__(cls, name, bases, dict_):
        
        # Let GObjectMeta do it's initialization
        GObjectMeta.__init__(cls, name, bases, dict_)
        
        do_drag_data_get = dict_.get('do_drag_data_get')
        if do_drag_data_get:
            
            repo = GIRepository.Repository.get_default()
            
            for base in cls.__mro__:
                typeinfo = repo.find_by_gtype(base.__gtype__)
                if typeinfo:
                    vfunc = GIRepository.object_info_find_vfunc_using_interfaces(typeinfo, 'drag_data_get')
                    if vfunc:
                        break
            else:
                raise AttributeError("Could not find vfunc for drag_data_get")
            
            # Get the address of the vfunc so we can put our own callback in there
            address = vfunc_info_get_address(vfunc[0], cls.__gtype__)
            if address == 0:
                raise AttributeError("Could not get address for drag_data_get")
            
            # Make a thunk function closure, store it so it doesn't go out of scope
            do_drag_data_get._thunk = create_drag_data_get_thunk(cls)
            
            # Don't judge me... couldn't get a normal function pointer to work
            dbl_pointer = ctypes.POINTER(ctypes.c_void_p)
            addr = ctypes.cast(address, dbl_pointer)
            addr.contents.value = ctypes.cast(do_drag_data_get._thunk, ctypes.c_void_p).value

class PatchedListStore(Gtk.ListStore):
    '''ListStore object that can be used with the high-level TreeView DnD API'''
    __metaclass__ = TreeDragSourceMeta

class PatchedTreeStore(Gtk.TreeStore):
    '''TreeStore object that can be used with the high-level TreeView DnD API'''
    __metaclass__ = TreeDragSourceMeta
